docs/source/environment.rst
Environment Variables
=====================
PyFarm's master and models have several environment variables which can
be used to change the operation at runtime.  For more information see the
individual sections below.

.. note::
    Not all environment variables defined below are directly used by
    PyFarm.  Many of these values are provided to make it easier to group
    settings together and so settings for PyFarm won't conflict with any
    existing software.

Database Schema
---------------
Environment variables that are used to setup or control the database backend.

.. warning::
    These values are used to construct the database schema.  If your schema
    already exists then changing these values may have uninteded consequences.


.. envvar:: PYFARM_DB_PREFIX

    The prefix for all table names.  Normally this should never be changed but
    could be for testing or similiar similiar circumstances.
    **NOTE**: making this value to long may produce errors in some databases
    such as MySQL

.. envvar:: PYFARM_DB_MAX_HOSTNAME_LENGTH

    The maximum length a hostname can be
    **Default**: 255

.. envvar:: PYFARM_DB_MAX_JOBTYPE_LENGTH

    The maximum length for the name of a jobtype.
    **Default**: 64

.. envvar:: PYFARM_DB_MAX_COMMAND_LENGTH

    The maximum length a single command can be.
    **Default**: 64

.. envvar:: PYFARM_DB_MAX_USERNAME_LENGTH

    The maximum length a username can be.
    **Default**: 255

.. envvar:: PYFARM_DB_MAX_EMAILADDR_LENGTH

    The maximum length a email address can be
    **Default**: 255

.. envvar:: PYFARM_DB_MAX_ROLE_LENGTH

    The maximum length a role can be
    **Default**: 128

.. envvar:: PYFARM_DB_MAX_TAG_LENGTH

    The maximum length a tag can be
    **Default**: 64

    **NOTE** PyFarm uses the word 'tag' in several places.  This value controls
    the max length of any string which is a tag.

.. envvar:: PYFARM_DB_MAX_PROJECT_NAME_LENGTH

    The maximum length any one project name can be.
    **Default**: 32

Database Constraints and Validation
-----------------------------------
Unlike the above section, these values are checked when a database entry is
modified or created.  They are intended to provide validation so erroneous
data cannot be inserted.  Do note however the **max** value any integer can
be raised to is 2147483647.

.. envvar:: PYFARM_AGENT_CPU_ALLOCATION

    The total amount of cpu space an agent is allowed to work in.  For example
    if four jobs requires four cpus and :envvar:`PYFARM_AGENT_CPU_ALLOCATION` is
    1.0 then all those jobs can be assigned to the agent. If
    :envvar:`PYFARM_AGENT_CPU_ALLOCATION` was .5 however only half of those jobs
    could be assigned.  This value must always be greater than 0.
    **Default**: .8

.. envvar:: PYFARM_AGENT_RAM_ALLOCATION

    Same as :envvar:`PYFARM_AGENT_CPU_ALLOCATION` except for ram resources.
    This value must always be greater than 0.
    **Default**: 1.0

.. envvar:: PYFARM_AGENT_MIN_PORT

    The minimum port an agent is allowed to communicate on.
    **Default**: 1024


.. envvar:: PYFARM_AGENT_MAX_PORT

    The maximum port an agent is allowed to communicate on.
    **Default**: 65535

.. envvar:: PYFARM_AGENT_MIN_CPUS

    The minimum number of cpus an agent is allowed to have.
    **Default**: 1

.. envvar:: PYFARM_AGENT_MAX_CPUS

    The maximum number of cpus an agent is allowed to have.
    **Default**: 256

.. envvar:: PYFARM_AGENT_MIN_RAM

    The minimum amount of ram, in megabytes, an agent is allowed to have.
    **Default**: 16

.. envvar:: PYFARM_AGENT_MAX_RAM

    The maximum amount of ram, in megabytes, an agent is allowed to have.
    **Default**: 262144

.. envvar:: PYFARM_QUEUE_MIN_PRIORITY

    The minimum priority any job or task is allowed to have.
    **Default**: -1000

.. envvar:: PYFARM_QUEUE_MAX_PRIORITY

    The maximum priority any job or task is allowed to have.
    **Default**: 1000

.. envvar:: PYFARM_QUEUE_DEFAULT_PRIORITY

    The default priority any new jobs or tasks are given
    **Default**: 0

.. envvar:: PYFARM_QUEUE_MIN_BATCH

    The minimum number of tasks which can be sent to a single agent for
    processing.
    **Default**: 1

.. envvar:: PYFARM_QUEUE_MAX_BATCH

    The maximum number of tasks which can be sent to a single agent for
    processing.
    **Default**: 64

.. envvar:: PYFARM_QUEUE_DEFAULT_BATCH

    The default number of tasks which can be sent to a single agent for
    processing.
    **Default**: 1

.. envvar:: PYFARM_QUEUE_MIN_REQUEUE

    The minimum number of times a task is allowed to reque.
    **Default**: 0

.. envvar:: PYFARM_QUEUE_MAX_REQUEUE

    The maximum number of times a task is allowed to reque.  Not setting this
    value will allow **any** tasks to reque an infinite number of times if
    requested by a user.
    **Default**: 10

.. envvar:: PYFARM_QUEUE_DEFAULT_REQUEUE

    The default number of times a task is allowed to reque.
    **Default**: 3

.. envvar:: PYFARM_QUEUE_MIN_CPUS

    The minimum number of cpus that can be required to any one job.
    **Default**: 1

.. envvar:: PYFARM_QUEUE_MAX_CPUS

    The maximum number of cpus that can be required to any one job.
    **Default**: 256

.. envvar:: PYFARM_QUEUE_DEFAULT_CPUS

    The default number of cpus required for any one job.
    **Default**: 1

.. envvar:: PYFARM_QUEUE_MIN_RAM

    The minimum amount of ram, in megabytes, that can be required for any one
    job.
    **Default**: 16

.. envvar:: PYFARM_QUEUE_MAX_RAM

    The maximum number of cpus that can be required to any one job.
    **Default**: 256

.. envvar:: PYFARM_QUEUE_DEFAULT_RAM

    The default amount of ram, in megabytes, that is required for a job.
    **Default**: 32

.. envvar:: PYFARM_REQUIRE_PRIVATE_IP

    Whether pyfarm-master should reject agents with non-private IP addresses
    **Default**: False

Master
------
Environment variables that are used within the server processes on the
master.

.. envvar:: PYFARM_CONFIG

    Controls which configuration should be loaded.  Currently the only
    supported values are `debug` and `prod` and the configuration itself
    is handled internally.

.. envvar:: PYFARM_DATABASE_URI

    The URI to connect to the backend database.  This should be a valid
    `sqlalchemy uri <http://docs.sqlalchemy.org/en/rel_0_8/core/engines.html#database-urls>`_
    which looks something like this::

        dialect+driver://user:password@host/dbname[?key=value..]

.. envvar:: PYFARM_SECRET_KEY

    When present this value is used by forms and the password storage as
    a seed value for several operations.

.. envvar:: PYFARM_CSRF_SESSION_KEY

    Key used to set the cross site request forgery key for use
    by :mod:`wtforms`.  If not provided this will be set to
    :envvar:`PYFARM_SECRET_KEY`

.. envvar:: PYFARM_JSON_PRETTY

    If set to `true` then all json output by the REST api will be human
    readable.  Setting :envvar:`PYFARM_CONFIG` to `debug` will also produce
    the same effect.

.. envvar:: PYFARM_API_VERSION

    The version of the REST api used for varying points of logic and
    for constructing :envvar:`PYFARM_API_PREFIX`

.. envvar:: PYFARM_API_PREFIX

    If set, this will establish the prefix for mounting the API.  This value
    is combined with :envvar:`PYFARM_API_VERSION` resulting in something along
    the lines of::

        https://$hostname/$PYFARM_API_PREFIX$PYFARM_API_VERSION

.. envvar:: JOBTYPE_DEFAULT_MAX_BATCH

    Performs the same function as :envvar:`PYFARM_QUEUE_MAX_BATCH` but provides
    an override specifically for :attr:`pyfarm.models.jobtype.JobType.max_batch`

.. envvar:: JOBTYPE_DEFAULT_BATCH_CONTIGUOUS

    Sets the default value for
    :attr:`pyfarm.models.jobtype.JobType.batch_contiguous`


pyfarm/models/core/cfg.py
# No shebang line, this module is meant to be imported
#
# Copyright 2013 Oliver Palmer
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Configuration Variables
=======================

Stores basic configuration data related to tables and models.  Most of these
variables have defaults defined in the configuration under `PYFARM_DB_<value>`

:const string TABLE_PREFIX:
    Prefix for all tables

:const string TABLE_SOFTWARE:
    Stores the name of the table for software items

:const string TABLE_TAG:
    Stores the name of the table for tags

:const string TABLE_AGENT:
    Stores the name of the table for agents

:const string TABLE_AGENT_TAGS:
    Stores the name of the table for agent tags

:const string TABLE_JOB:
    Stores the name of the table for jobs

:const string TABLE_JOB_TAG:
    Stores the name of the table for job tags

:const string TABLE_TASK:
    Stores the name of the table for job tasks

:const string TABLE_USERS_USER:
    Stores the registered users (both human and api)

:const string TABLE_USERS_ROLE:
    Stores roles in which a user can operate in

:const string TABLE_USERS_USER_ROLE:
    Stores relationships between :const:`.TABLE_USERS_USER` and
    :const:`.TABLE_USERS_ROLE`

:const string TABLE_JOB_QUEUES:
    Stores the name of the table for job queues

:const integer MAX_HOSTNAME_LENGTH:
    the max length of a hostname

:const integer MAX_JOBTYPE_LENGTH:
    the max length of a jobtype

:const integer MAX_COMMAND_LENGTH:
    the max length of a command (ex. `bash` or `cmd.exe`)

:const integer MAX_USERNAME_LENGTH:
    the max length of a username

:const integer MAX_TAG_LENGTH:
    the max length of a tag

    .. note::
        this value is shared amongst all tag columns and may be split into
        multiple values at a later time
"""

from pyfarm.core.config import read_env, read_env_int

# table names
TABLE_PREFIX = read_env("PYFARM_DB_PREFIX", "pyfarm_")
TABLE_SOFTWARE = "%ssoftware" % TABLE_PREFIX
TABLE_SOFTWARE_VERSION = "%s_version" % TABLE_SOFTWARE
TABLE_TAG = "%stag" % TABLE_PREFIX
TABLE_AGENT = "%sagents" % TABLE_PREFIX
TABLE_AGENT_SOFTWARE_VERSION_ASSOC = "%s_software_version_assoc" % TABLE_AGENT
TABLE_AGENT_TAG_ASSOC = "%s_tag_assoc" % TABLE_AGENT
TABLE_JOB = "%sjobs" % TABLE_PREFIX
TABLE_JOB_TYPE = "%s_jobtypes" % TABLE_JOB
TABLE_JOB_TYPE_VERSION = "%s_versions" % TABLE_JOB_TYPE
TABLE_JOB_TAG_ASSOC = "%s_tag_assoc" % TABLE_JOB
TABLE_JOB_DEPENDENCIES = "%s_dependencies" % TABLE_JOB
TABLE_JOB_SOFTWARE_REQ = "%s_software_req" % TABLE_JOB
TABLE_JOB_NOTIFIED_USERS = "%s_notified_users" % TABLE_PREFIX
TABLE_JOB_TYPE_SOFTWARE_REQ = "%sjobtype_software_req" % TABLE_PREFIX
TABLE_TASK = "%stask" % TABLE_PREFIX
TABLE_TASK_DEPENDENCIES = "%s_dependencies" % TABLE_TASK
TABLE_USERS = "%susers" % TABLE_PREFIX
TABLE_USERS_PROJECTS = "%s_projects" % TABLE_USERS
TABLE_USERS_USER = "%s_users" % TABLE_USERS
TABLE_USERS_ROLE = "%s_roles" % TABLE_USERS
TABLE_USERS_USER_ROLES = "%s_user_roles" % TABLE_USERS
TABLE_PROJECT = "%sprojects" % TABLE_PREFIX
TABLE_PROJECT_AGENTS = "%s_agents" % TABLE_PROJECT
TABLE_JOB_QUEUE = "%sjob_queues" % TABLE_PREFIX

TABLES = (TABLE_SOFTWARE, TABLE_SOFTWARE_VERSION, TABLE_TAG,
          TABLE_AGENT_SOFTWARE_VERSION_ASSOC, TABLE_AGENT, TABLE_JOB_TYPE,
          TABLE_AGENT_TAG_ASSOC, TABLE_USERS_USER, TABLE_USERS_ROLE,
          TABLE_USERS_USER_ROLES, TABLE_TASK, TABLE_TASK_DEPENDENCIES,
          TABLE_JOB_DEPENDENCIES, TABLE_JOB_TAG_ASSOC, TABLE_JOB_SOFTWARE_REQ,
          TABLE_JOB_NOTIFIED_USERS, TABLE_JOB_TYPE_SOFTWARE_REQ, TABLE_JOB,
          TABLE_PROJECT, TABLE_PROJECT_AGENTS, TABLE_USERS_PROJECTS,
          TABLE_JOB_TYPE_VERSION, TABLE_JOB_QUEUE)

# column lengths
MAX_HOSTNAME_LENGTH = read_env_int("PYFARM_DB_MAX_HOSTNANE_LENGTH", 255)
MAX_JOBTITLE_LENGTH = read_env_int("PYFARM_DB_MAX_JOBTTITLE_LENGTH", 255)
MAX_JOBTYPE_LENGTH = read_env_int("PYFARM_DB_MAX_JOBTYPE_LENGTH", 64)
MAX_COMMAND_LENGTH = read_env_int("PYFARM_DB_MAX_COMMAND_LENGTH", 64)
MAX_USERNAME_LENGTH = read_env_int("PYFARM_DB_MAX_USERNAME_LENGTH", 255)
MAX_EMAILADDR_LENGTH = read_env_int("PYFARM_DB_MAX_EMAILADDR_LENGTH", 255)
SHA256_ASCII_LENGTH = 64  # static length of a sha256 string
MAX_ROLE_LENGTH = read_env_int("PYFARM_DB_MAX_ROLE_LENGTH", 128)
MAX_TAG_LENGTH = read_env_int("PYFARM_DB_MAX_TAG_LENGTH", 64)
MAX_PROJECT_NAME_LENGTH = read_env_int("PYFARM_DB_MAX_PROJECT_NAME_LENGTH", 32)
MAX_JOBQUEUE_NAME_LENGTH = read_env_int("PYFARM_MAX_JOBQUEUE_NAME_LENGTH", 255)
